Fix rbenv module losing build_env config

brlanier opened the pull request:
### What does this PR do?
Fixes the rbenv module issue with not passing the build_config compile environment options to the command line for use. They were getting lost along the way. At some point in the past a change was made on how the options where being handled as strings and I believe that is when this stopped working. I don't know that many people use this option and my use case was trivial so I never bothered to debug. Please see the referenced issue for more details on the problem.

Originally, the build_env was passed as a string, converted to a list, back to a string and then expected to be a list or dictionary. If it was a list, it was converted to a dictionary. This PR changes the expectation of the pillar/grain input to be in a dictionary format, while retaining compatibility for previous setups. 

Also, not a python expert so guessing on some of the use of the built in tools and practices I worked with. 

### What issues does this PR fix or reference?
Fixes #31022 

### Previous Behavior
The rbenv state would work fine but any compile time options you might have tried to set via pillar or grains would be ignored. For instance, if you tried to compile a ruby with `RUBY_CONFIGURE_OPTS="--with-jemalloc`, that compile time option never made it to the ruby build process.

### New Behavior
The build_env pillar/grains options are now handled properly and not lost in the process. If you use the state to compile a ruby with extra options, you will now see the compiled in options present.

### Tests written?
Tests updated. I know enough python to get this far, so the tests could use some review

I did test this works for my old configuration that was previously getting silently ignored and I tested under the new dictionary setup I propose for going forward since this appears to be what is asked for in the code anyway.

### Commits signed with GPG?

No

brlanier committed:
Fix rbenv module losing build_env config

Fixes #31022

- Refactor rbenv module build_env compile options to use a dictionary
  via pillar/grains/master config.
- Was converting string to list back to string and then ignoring when it
  wasn't a list or dictionary.
- Should be backwards compatible with current config setups while
  promoting new style in docs, although the old string style config
  hasn't worked in a long time.

brlanier commented:
Updated version of #55383 based on master

dwoz committed:
Merge branch 'master' into 31022-fix-rbenv-build-env-parsingv2

dwoz reviewed (changes requested):
@brlanier Looks like the docs build had problems:
```
03:32:13  Warning, treated as error:
03:32:13  /var/jenkins/workspace/pr-docs_PR-55496/salt/states/rbenv.py:docstring of salt.states.rbenv:24:Unexpected indentation.
03:32:15  Makefile:56: recipe for target 'html' failed
03:32:15  make: *** [html] Error 2
```

brlanier committed:
Fix rbenv state documentation formatting

brlanier committed:
Merge branch '31022-fix-rbenv-build-env-parsingv2' of github.com:amco/salt into 31022-fix-rbenv-build-env-parsingv2

dwoz reviewed (changes requested):
@brlanier Sorry I didn't catch this before but the test failures seem to be related to this change.

brlanier commented on salt/modules/rbenv.py:
```diff
@@ -232,19 +234,21 @@ def install_ruby(ruby, runas=None):
+    if isinstance(build_env, dict):
+        env.update(build_env)
+    elif isinstance(build_env, str):
+        env.update(dict(build_arg.split('=', 1) for build_arg in salt.utils.args.shlex_split(six.text_type(build_env))))
```
Looks like the python 3 test/builds are failing on this line. Will have to try to find time to dig into why. I figured this line might be an issue with both versions and I was guessing on trying to use some of the builtins to help with that.

brlanier replied:
okay, it errors when build_env is empty for some reason only in the python 3 builds but not python 2.
In the test suite we have
```
with patch.dict(rbenv.__salt__,
                            {'config.get': MagicMock(return_value='True')}):
```
Which if I understand correctly is just returning config.get with nothing.. just saying it worked. Either way, classic mistake of me testing with the condition I wanted but not with the condition that could exist of not wanting compile options. Will work on a fix. 
I suppose, another test should be added to test when config.get does return a build option, but I barely understand what I am writing here let alone test syntax.

brlanier replied:
The issue was in the split when the test value for the config was just `True`.  There was no `=` to split on thus not creating the proper number of args for the dict build process.

brlanier committed:
Merge branch 'master' into 31022-fix-rbenv-build-env-parsingv2

brlanier committed:
Fix issue with tests

- Was failing on split when no '=' in the build_env config
- Added some additional tests to check the different config types

brlanier commented on tests/unit/modules/test_rbenv.py:
```diff
@@ -76,6 +76,49 @@ def test_install_ruby(self):
+                                      return_value=None):
+                        self.assertFalse(rbenv.install_ruby('ruby'))
+
+
```
I have pretty much zero idea what I am doing in here. My intention was to test the different config options the install_ruby function could handle, along with the default.  These tests past on my local environment, but I don't know if there is a better way to do what I was trying to do here. So I just copied the original test and changed the mock data to fit the different kinds of config options the method could take.
Open to suggestions.

brlanier committed:
Fix spacing on lint issue

brlanier committed:
Merge branch 'master' into 31022-fix-rbenv-build-env-parsingv2

brlanier commented:
@dwoz I don't believe these recent test failures are related to the changes I made after updating the tests. Not sure why those four are errored, if it is these changes or other changes in salt at the moment.

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.
